File: fud-ai/ios/calorietracker/Services/WeightAnalysisService.swift
import Foundation

/// Pure thermodynamic / statistical forecast of where the user's weight is heading based on
/// recent calorie intake, logged weight history, and their profile (gender, age, activity, goal).
/// No network, no LLM — just energy-balance math + linear regression.
struct WeightForecast {
    /// Average daily calories eaten over the last `lookbackDays` days of logged food.
    let avgDailyCalories: Int
    /// Total Daily Energy Expenditure from the user's profile (BMR × activity).
    let tdee: Int
    /// avgDailyCalories − tdee. Negative = deficit, positive = surplus.
    let dailyEnergyBalance: Int

    /// Predicted kg change per week based on energy balance (7,700 kcal ≈ 1 kg body fat).
    let predictedWeeklyChangeKg: Double
    /// Predicted kg change per week derived from linear regression on recent weight log.
    /// Nil if fewer than 2 weight entries in the regression window.
    let observedWeeklyChangeKg: Double?

    /// Current weight (latest entry in the store, or profile weight if no entries).
    let currentWeightKg: Double
    /// Predicted weight 30 days from now at the current predicted pace.
    let predictedWeight30dKg: Double
    let predictedWeight60dKg: Double
    let predictedWeight90dKg: Double

    /// Days until goal weight at the current predicted pace. Nil if no goal, if goal is Maintain,
    /// or if the user is moving in the wrong direction.
    let daysToGoal: Int?
    let goalReachDate: Date?

    /// Sentinel for views to know whether enough data exists to render a meaningful forecast.
    let hasEnoughData: Bool
    /// True when predicted (from diet) and observed (from weight scale) trends disagree substantially.
    /// Hint that the user may be under-logging food or not weighing consistently.
    let trendsDisagree: Bool

    /// How many days of logged food informed the calorie average. Displayed so the user knows
    /// how much history the forecast is based on.
    let daysOfFoodData: Int
    /// How many weight entries informed the observed-trend regression.
    let weightEntriesUsed: Int

    /// Hard cap so ancient data doesn't dominate the projection. Everything within this window
    /// is used — if the user only has a week, that's fine; if they have 3 months, we use 3 months.
    static let maxLookbackDays: Int = 90
}

enum WeightAnalysisService {
    static func compute(weights: [WeightEntry], foods: [FoodEntry], profile: UserProfile) -> WeightForecast {
        let now = Date()
        let calendar = Calendar.current
        let cutoff = calendar.date(byAdding: .day, value: -WeightForecast.maxLookbackDays, to: now) ?? now

        // --- Avg daily calories from food log (up to 90 days of available data) ---
        let recentFoods = foods.filter { $0.timestamp >= cutoff && $0.timestamp <= now }
        let daysLogged = Set(recentFoods.map { calendar.startOfDay(for: $0.timestamp) }).count
        let totalRecentCal = recentFoods.reduce(0) { $0 + $1.calories }
        let avgDailyCal = daysLogged > 0 ? totalRecentCal / daysLogged : 0

        let tdee = Int(profile.tdee)
        let balance = avgDailyCal - tdee
        // 7,700 kcal ≈ 1 kg body fat (ISSN standard for deficit/surplus math).
        let predictedWeeklyKg = Double(balance) * 7.0 / 7_700.0

        // --- Current weight ---
        let sortedWeights = weights.sorted { $0.date > $1.date }
        let currentWeight = sortedWeights.first?.weightKg ?? profile.weightKg

        // --- Observed trend via linear regression on weight history ---
        let regressionWindow = sortedWeights.filter { $0.date >= cutoff }
        let observedWeeklyKg = linearRegressionSlopePerDay(entries: regressionWindow).map { $0 * 7.0 }

        // --- Future projections ---
        let pred30 = currentWeight + predictedWeeklyKg * 30.0 / 7.0
        let pred60 = currentWeight + predictedWeeklyKg * 60.0 / 7.0
        let pred90 = currentWeight + predictedWeeklyKg * 90.0 / 7.0

        // --- Goal-reach projection ---
        var daysToGoal: Int? = nil
        var goalReachDate: Date? = nil
        if let goalKg = profile.goalWeightKg, predictedWeeklyKg != 0, profile.goal != .maintain {
            let kgRemaining = goalKg - currentWeight
            let movingCorrectWay = (profile.goal == .lose && predictedWeeklyKg < 0 && kgRemaining < 0)
                                || (profile.goal == .gain && predictedWeeklyKg > 0 && kgRemaining > 0)
            if movingCorrectWay {
                let daysPerKg = 7.0 / abs(predictedWeeklyKg)
                let days = Int((abs(kgRemaining) * daysPerKg).rounded())
                daysToGoal = days
                goalReachDate = calendar.date(byAdding: .day, value: days, to: now)
            }
        }

        // Work with whatever the user has. Even 2 days of food and 2 weights is enough for a rough forecast;
        // the LLM will caveat accordingly.
        let hasEnoughData = daysLogged >= 2 && weights.count >= 2

        // Compare predicted vs observed trends. If both exist and they differ by more than
        // 0.3 kg/week, user is likely under-logging food (or over-eating un-logged snacks).
        let trendsDisagree: Bool
        if let observed = observedWeeklyKg, hasEnoughData {
            trendsDisagree = abs(predictedWeeklyKg - observed) > 0.3
        } else {
            trendsDisagree = false
        }

        return WeightForecast(
            avgDailyCalories: avgDailyCal,
            tdee: tdee,
            dailyEnergyBalance: balance,
            predictedWeeklyChangeKg: predictedWeeklyKg,
            observedWeeklyChangeKg: observedWeeklyKg,
            currentWeightKg: currentWeight,
            predictedWeight30dKg: pred30,
            predictedWeight60dKg: pred60,
            predictedWeight90dKg: pred90,
            daysToGoal: daysToGoal,
            goalReachDate: goalReachDate,
            hasEnoughData: hasEnoughData,
            trendsDisagree: trendsDisagree,
            daysOfFoodData: daysLogged,
            weightEntriesUsed: regressionWindow.count
        )
    }

    /// Slope of a simple linear regression (y = mx + b) over weight entries, returning m in
    /// kg per day. Returns nil if fewer than 2 entries.
    private static func linearRegressionSlopePerDay(entries: [WeightEntry]) -> Double? {
        guard entries.count >= 2 else { return nil }
        // Use seconds-since-epoch as X so slope has real time units, then scale to per-day.
        let xs = entries.map { $0.date.timeIntervalSince1970 }
        let ys = entries.map { $0.weightKg }
        let n = Double(xs.count)
        let meanX = xs.reduce(0, +) / n
        let meanY = ys.reduce(0, +) / n
        var num = 0.0
        var den = 0.0
        for i in 0..<xs.count {
            let dx = xs[i] - meanX
            num += dx * (ys[i] - meanY)
            den += dx * dx
        }
        guard den != 0 else { return nil }
        let kgPerSecond = num / den
        return kgPerSecond * 86_400.0
    }
}

private extension Comparable {
    func clamped(to limits: ClosedRange<Self>) -> Self {
        min(max(self, limits.lowerBound), limits.upperBound)
    }
}

File: fud-ai/android/app/src/main/java/com/apoorvdarshan/calorietracker/services/WeightAnalysisService.kt
package com.apoorvdarshan.calorietracker.services

import com.apoorvdarshan.calorietracker.models.FoodEntry
import com.apoorvdarshan.calorietracker.models.UserProfile
import com.apoorvdarshan.calorietracker.models.WeightEntry
import com.apoorvdarshan.calorietracker.models.WeightGoal
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * Pure thermodynamic / statistical forecast of where the user's weight is heading based on
 * recent calorie intake, logged weight history, and their profile. No network, no LLM —
 * energy-balance math + linear regression.
 */
data class WeightForecast(
    val avgDailyCalories: Int,
    val tdee: Int,
    val dailyEnergyBalance: Int,
    val predictedWeeklyChangeKg: Double,
    val observedWeeklyChangeKg: Double?,
    val currentWeightKg: Double,
    val predictedWeight30dKg: Double,
    val predictedWeight60dKg: Double,
    val predictedWeight90dKg: Double,
    val daysToGoal: Int?,
    val goalReachDate: Instant?,
    val hasEnoughData: Boolean,
    val trendsDisagree: Boolean,
    val daysOfFoodData: Int,
    val weightEntriesUsed: Int
) {
    companion object {
        const val MAX_LOOKBACK_DAYS = 90
    }
}

data class AdaptiveGoalResult(
    val profile: UserProfile,
    val changed: Boolean,
    val updatedCalories: Int?,
    val message: String
)

object AdaptiveGoalService {
    private const val MINIMUM_FOOD_DAYS = 4
    private const val MINIMUM_WEIGHT_ENTRIES = 3
    private const val MINIMUM_DAILY_ADJUSTMENT = 25
    private const val MAXIMUM_DAILY_ADJUSTMENT = 150
    private const val CALORIES_PER_KG = 7_700.0

    /**
     * [measuredTdee] (Health Connect active + basal energy, kcal/day) is preferred over the
     * formula TDEE when available: it sets the safety ceiling and, when there isn't yet enough
     * weight-trend data, drives a burn-based correction toward measured maintenance + the goal
     * pace. The weight-trend correction stays primary whenever enough logs/weigh-ins exist.
     */
    fun apply(
        profile: UserProfile,
        weights: List<WeightEntry>,
        foods: List<FoodEntry>,
        measuredTdee: Int? = null
    ): AdaptiveGoalResult {
        val forecast = WeightAnalysisService.compute(weights = weights, foods = foods, profile = profile)
        val observedWeeklyChangeKg = forecast.observedWeeklyChangeKg
        val targetWeeklyChangeKg = targetWeeklyChangeKg(profile)
        val currentCalories = profile.effectiveCalories
        val safetyFloor = max(profile.bmr.roundToInt(), 1_200)
        // Prefer Health Connect measured burn for the maintenance/ceiling basis when available.
        val maintenanceTdee = measuredTdee ?: profile.tdee.roundToInt()
        val safetyCeiling = max(safetyFloor, (maintenanceTdee * 1.25).roundToInt())

        val hasWeightTrend = forecast.daysOfFoodData >= MINIMUM_FOOD_DAYS &&
            forecast.weightEntriesUsed >= MINIMUM_WEIGHT_ENTRIES &&
            observedWeeklyChangeKg != null

        val limitedAdjustment: Int = if (hasWeightTrend && observedWeeklyChangeKg != null) {
            // Primary: correct from the real weight trend vs the target pace.
            val raw = (targetWeeklyChangeKg - observedWeeklyChangeKg) * CALORIES_PER_KG / 7.0
            raw.roundToInt().coerceIn(-MAXIMUM_DAILY_ADJUSTMENT, MAXIMUM_DAILY_ADJUSTMENT)
        } else if (measuredTdee != null) {
            // Not enough weigh-ins/food yet, but Health Connect gives measured burn: steer the
            // target toward measured maintenance + the goal pace.
            val targetCalories = measuredTdee + (targetWeeklyChangeKg * CALORIES_PER_KG / 7.0).roundToInt()
            (targetCalories - currentCalories).coerceIn(-MAXIMUM_DAILY_ADJUSTMENT, MAXIMUM_DAILY_ADJUSTMENT)
        } else {
            return AdaptiveGoalResult(
                profile = profile,
                changed = false,
                updatedCalories = null,
                message = "Adaptive Goals is on. It needs at least $MINIMUM_FOOD_DAYS logged food days and $MINIMUM_WEIGHT_ENTRIES recent weight entries — or Health Connect energy data — before making a correction."
            )
        }

        if (abs(limitedAdjustment) < MINIMUM_DAILY_ADJUSTMENT) {
            return AdaptiveGoalResult(
                profile = profile,
                changed = false,
                updatedCalories = null,
                message = "Your recent trend is close to your selected goal pace, so Adaptive Goals did not change calories this week."
            )
        }

        if (limitedAdjustment < 0 && currentCalories <= safetyFloor) {
            return AdaptiveGoalResult(
                profile = profile,
                changed = false,
                updatedCalories = null,
                message = "Adaptive Goals did not lower calories because your current target is already at the safety floor."
            )
        }
        if (limitedAdjustment > 0 && currentCalories >= safetyCeiling) {
            return AdaptiveGoalResult(
                profile = profile,
                changed = false,
                updatedCalories = null,
                message = "Adaptive Goals did not raise calories because your current target is already at the safety ceiling."
            )
        }

        val proposedCalories = currentCalories + limitedAdjustment
        val adjustedCalories = if (limitedAdjustment < 0) {
            max(proposedCalories, safetyFloor)
        } else {
            min(proposedCalories, safetyCeiling)
        }

        if (adjustedCalories == currentCalories) {
            return AdaptiveGoalResult(
                profile = profile,
                changed = false,
                updatedCalories = null,
                message = "Adaptive Goals checked your trend, but calorie guardrails kept this week's target unchanged."
            )
        }

        val nextProfile = profile.copy(customCalories = adjustedCalories)
        val signedAdjustment = adjustedCalories - currentCalories
        val sign = if (signedAdjustment > 0) "+" else ""
        val basis = if (hasWeightTrend) "your recent weight trend" else "your Health Connect energy burn"
        return AdaptiveGoalResult(
            profile = nextProfile,
            changed = true,
            updatedCalories = adjustedCalories,
            message = "Adaptive Goals adjusted calories by $sign$signedAdjustment kcal to $adjustedCalories kcal based on $basis. Pinned macros stay pinned; unlocked macros auto-balance."
        )
    }

    private fun targetWeeklyChangeKg(profile: UserProfile): Double = when (profile.goal) {
        WeightGoal.LOSE -> -(profile.weeklyChangeKg ?: 0.5)
        WeightGoal.MAINTAIN -> 0.0
        WeightGoal.GAIN -> profile.weeklyChangeKg ?: 0.5
    }
}

object WeightAnalysisService {

    fun compute(
        weights: List<WeightEntry>,
        foods: List<FoodEntry>,
        profile: UserProfile
    ): WeightForecast {
        val now = Instant.now()
        val zone = ZoneId.systemDefault()
        val cutoff = now.minusSeconds(WeightForecast.MAX_LOOKBACK_DAYS * 86_400L)

        val recentFoods = foods.filter { it.timestamp in cutoff..now }
        val daysLogged = recentFoods.map { it.timestamp.atZone(zone).toLocalDate() }.toSet().size
        val totalRecentCal = recentFoods.sumOf { it.calories }
        val avgDailyCal = if (daysLogged > 0) totalRecentCal / daysLogged else 0

        val tdee = profile.tdee.toInt()
        val balance = avgDailyCal - tdee
        // 7,700 kcal ≈ 1 kg body fat (ISSN standard for deficit/surplus math).
        val predictedWeeklyKg = balance.toDouble() * 7.0 / 7_700.0

        val sortedWeights = weights.sortedByDescending { it.date }
        val currentWeight = sortedWeights.firstOrNull()?.weightKg ?: profile.weightKg

        val regressionWindow = sortedWeights.filter { it.date >= cutoff }
        val observedWeeklyKg = linearRegressionSlopePerDay(regressionWindow)?.let { it * 7.0 }

        val pred30 = currentWeight + predictedWeeklyKg * 30.0 / 7.0
        val pred60 = currentWeight + predictedWeeklyKg * 60.0 / 7.0
        val pred90 = currentWeight + predictedWeeklyKg * 90.0 / 7.0

        var daysToGoal: Int? = null
        var goalReachDate: Instant? = null
        val goalKg = profile.goalWeightKg
        if (goalKg != null && predictedWeeklyKg != 0.0 && profile.goal != WeightGoal.MAINTAIN) {
            val kgRemaining = goalKg - currentWeight
            val movingCorrectWay =
                (profile.goal == WeightGoal.LOSE && predictedWeeklyKg < 0 && kgRemaining < 0) ||
                        (profile.goal == WeightGoal.GAIN && predictedWeeklyKg > 0 && kgRemaining > 0)
            if (movingCorrectWay) {
                val daysPerKg = 7.0 / abs(predictedWeeklyKg)
                val days = (abs(kgRemaining) * daysPerKg).roundToInt()
                daysToGoal = days
                goalReachDate = now.plusSeconds(days * 86_400L)
            }
        }

        val hasEnoughData = daysLogged >= 2 && weights.size >= 2

        val trendsDisagree = observedWeeklyKg?.let { observed ->
            hasEnoughData && abs(predictedWeeklyKg - observed) > 0.3
        } ?: false

        return WeightForecast(
            avgDailyCalories = avgDailyCal,
            tdee = tdee,
            dailyEnergyBalance = balance,
            predictedWeeklyChangeKg = predictedWeeklyKg,
            observedWeeklyChangeKg = observedWeeklyKg,
            currentWeightKg = currentWeight,
            predictedWeight30dKg = pred30,
            predictedWeight60dKg = pred60,
            predictedWeight90dKg = pred90,
            daysToGoal = daysToGoal,
            goalReachDate = goalReachDate,
            hasEnoughData = hasEnoughData,
            trendsDisagree = trendsDisagree,
            daysOfFoodData = daysLogged,
            weightEntriesUsed = regressionWindow.size
        )
    }

    /**
     * Slope of a simple linear regression (y = mx + b) over weight entries, returning m in
     * kg per day. Returns null if fewer than 2 entries or all x's are the same.
     */
    private fun linearRegressionSlopePerDay(entries: List<WeightEntry>): Double? {
        if (entries.size < 2) return null
        val xs = entries.map { it.date.epochSecond.toDouble() }
        val ys = entries.map { it.weightKg }
        val n = xs.size.toDouble()
        val meanX = xs.sum() / n
        val meanY = ys.sum() / n
        var num = 0.0
        var den = 0.0
        for (i in xs.indices) {
            val dx = xs[i] - meanX
            num += dx * (ys[i] - meanY)
            den += dx * dx
        }
        if (den == 0.0) return null
        val kgPerSecond = num / den
        return kgPerSecond * 86_400.0
    }
}

@Suppress("unused")
private fun Instant.toLocalDateInZone(zone: ZoneId): LocalDate =
    this.atZone(zone).toLocalDate()
